types/mdx/index.d.ts

```diff
@@ -0,0 +1,60 @@
+// Type definitions for non-npm package mdx 2.0
+// Project: https://github.com/mdx-js/mdx
+// Definitions by: Christian Murphy <https://github.com/ChristianMurphy>
+//                 Remco Haszing <https://github.com/remcohaszing>
+//                 Titus Wormer <https://github.com/wooorm>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 4.1
+
+/**
+ * An MDX file which exports a JSX component.
+ *
+ * By default MDX files export a single function which takes props and returns a JSX element.
+ * However, MDX files may export more, either through ESM exports inside the MDX file, or using by
+ * exports generated by plugins.
+ *
+ * To define exports for every MDX file in a project, create a TypeScript
+ * [script](https://www.typescriptlang.org/docs/handbook/2/modules.html#non-modules) file which
+ * augments `*.mdx`. A script file is a file which doesn’t use top-level ESM syntax, but ESM syntax
+ * us allowed inside the declared module. This is typically useful for exports created by plugins.
+ * For example:
+ *
+ * ```ts
+ * // mdx-custom.d.ts
+ * declare module '*.mdx' {
+ *   import { Frontmatter } from 'my-frontmatter-types';
+ *
+ *   export const frontmatter: Frontmatter;
+ *   export const title: string;
+ * }
+ * ```
+ *
+ * To define additional types for specific MDX files, create a file named the same, but postfixed
+ * with `.d.ts` next to the MDX file.
+ *
+ * For example, given the following MDX file `my-component.mdx`:
+ *
+ * ```mdx
+ * export const hello = 'mdx';
+ *
+ * # Hello MDX
+ * ```
+ *
+ * Create the following file named `my-component.mdx.d.ts` in the same directory:
+ *
+ * ```ts
+ * export const hello: string;
+ * ```
+ *
+ * It should now be possible to import both the MDX component and the exported constant `hello`.
+ */
+declare module '*.mdx' {
+    /**
+     * An function component which renders the MDX content using JSX.
+     *
+     * @param props This value is be available as the named variable `props` inside the MDX component.
+     * @returns A JSX element. The meaning of this may depend on the project configuration. I.e. it
+     * could be a React, Preact, or Vuex element.
+     */
+    export default function MDXContent(props: Record<string, unknown>): JSX.Element;
```

Comment:
Should we make this smarter?

I tried to type it like this yesterday: https://github.com/wooorm/xdm/blob/db77de625e018257631543591d5794a2e88952bb/complex-types.d.ts#L54. The benefit is that `props.components` is typed.
The downside is that not every JSX runtime supports components / JSX.Element properly — Preact had some problems (React / Vue were fine).

Reply:
Does this suffice?

```ts
declare module '*.mdx' {
    type MDXComponentsProp = Record<
        string,
        JSX.ElementClass | keyof JSX.IntrinsicElements | ((props: Record<string, unknown>) => JSX.Element)
    >;
    type MDXProps = Record<string, unknown> & { components?: MDXComponentsProp };

    export default function MDXContent(props: MDXProps): JSX.Element;
}
```

Although this would be nice, `JSX.ElementClass` and `JSX.IntrinsicElements` don’t have to be supported by a JSX runtime. Can we detect if a type exists on a runtime?

I tried the following, but it says JSX can’t be used as a type:

```ts
JSX extends { ElementClass: infer T } ? T : never
```

I also tried the following, but it says JSX can’t be used as a value:

```ts
typeof JSX extends { ElementClass: infer T } ? T : never
```

Ultimately I believe not specifying `components` (meaning they’re `unknown`) is better than specifying them incorrectly.

---

> The downside is that not every JSX runtime supports components / JSX.Element properly

Could you explain what exactly isn’t supported? The `JSX.Element` type is key to using JSX at all. Frameworks can support `createElement.JSX.Element` as well (where `createElement` represents the JSX pragma function), but JSX supports the automatic runtime, which relies on the global `JSX` namespace.